### PR TITLE
Fix for EXT-X-KEY URI not handling relative URL (Issue #202)

### PR DESCRIPTION
**Pull request by stevemayhew**
Here's a proposed fix for handling the URI if it's relative to the playlist (on the same level as segment).

All it needs is some unit test cases, the existing test cases pass.

### CODE REVIEW

**Comment by stevemayhew** [comment]
The build is not failing with my last checkin?  Why does this message persist?  Is there something else I need to do.

**Comment by gkatsev** [comment]
Looks like there's a test error now https://travis-ci.org/videojs/videojs-contrib-hls/builds/47580832
The issue is that on travis we run phantomjs whereas locally, we use karma to run it in chrome.

**Comment by stevemayhew** [comment]
Ah, that is this one:


 <>Running "karma:phantomjs" (karma) task
 <>PhantomJS 1.9.8 (Linux) ERROR
 <>	TypeError: 'null' is not an object (evaluating 'toolbar.appendChild')
 <>	at /home/travis/build/videojs/videojs-contrib-hls/node_modules/qunitjs/qunit/qunit.js:2534

Looks like I need to fire up my linux VM?  We have had lots of trouble with PhantomJS tests, particularly on Windows.

Gary, 
I am working on a video/audio sync issue now too…  Will put up the bug and sample playlist shortly.
Can you please connect me the person who manages the VideoJS effort at Bightcove.  The company I am working with is in the end game of evaluating video playback tech for HLS solutions.  There are a couple of flash only solutions as well as VisualOn in the running.

I’m a VideoJS fan though, because:
1) It follows HTML5 (so it falls back to native Safari for example)
2) lack of need for plugins is huge for us (FlashPlayer is really part of the browser now)
3) Much nicer to work in JS then AS3 (IMNSHO)

Would love to contribute materially to the project, but my management needs to get the warm and fuzzy to narrow in the bet on the winning horse.

You have heard of the company, they are a household name, a bit of Googling and you’ll figure it out ;-).

PS hope you guys are digging out ok out there!

Best Regards,
Steve

Steve Mayhew
Skype: stevemayhew
Twitter: stevemayhew
Mobile: 650-619-2529

**Comment by stevemayhew** [comment]
Ah, that is this one:

 <>Running "karma:phantomjs" (karma) task
 <>PhantomJS 1.9.8 (Linux) ERROR
 <>	TypeError: 'null' is not an object (evaluating 'toolbar.appendChild')
 <>	at /home/travis/build/videojs/videojs-contrib-hls/node_modules/qunitjs/qunit/qunit.js:2534

Looks like I need to fire up my linux VM?  We have had lots of trouble with PhantomJS tests, particularly on Windows.

I'll try and see if I can reproduce it.

File: src/videojs-hls.js [hunk 1]
```diff
@@ -474,13 +474,20 @@ videojs.Hls.prototype.fillBuffer = function(offset) {
   }
 
   // resolve the segment URL relative to the playlist
+  segmentUri = this.resolveSegmentUrl(segment.uri);
+
+  this.loadSegment(segmentUri, offset);
+};
+
+videojs.Hls.prototype.resolveSegmentUrl = function(segmentRelativeUrl) {
```
Comment (dmlap) [inline src/videojs-hls.js]: I'm wondering if there's a better name for this method now that we'll be using it for more than just segment URL resolution.

Comment (stevemayhew) [inline src/videojs-hls.js]: Yeah, me too.  How about:
1. resolvePlaylistRelativeUri
2. playlistUriToUrl
3. resolveUrl (might be hard to search for vs similar names)

Personally I like option 2 best, we are resolving a URI in the playlist (which may be a URL) to a URL.  PS, for people that always confuse the two (like me ;-)).   Here's a simple non-geekly explanation of the difference that is pretty accurate [URL vs URI](https://danielmiessler.com/study/url_vs_uri/)

**Comment by dmlap** [comment]
Merged via 66510e67aad74474b6ca44fb98d580f1ee008c63.

@stevemayhew had a bit of free time so I added unit tests. Let me know if you don't think they cover the case we were mishandling properly. Thank you for the contribution!

**Pull request closed without merging**